File: src/chipschallenge/LevelFactory.java
package chipschallenge;

/**
 *
 * @author patrik
 */
public abstract class LevelFactory {
    public abstract GameLevel getLevel(int n);
    public abstract int getLastLevelNumber();
    public abstract int getLevelNumberByPassword(String pass);
    public final GameLevel getLevelByPassword(String pass) {
        int n = getLevelNumberByPassword(pass);
        if(n == -1)
            return null;
        return getLevel(n);
    }
}

File: ChipsChallenge.Shared/LevelFactory.cs
namespace ChipsChallenge.Shared
{
    public abstract class LevelFactory
    {
        public abstract int LastLevelNumber { get; }

        public GameLevel GetLevelByPassword(string password)
        {
            int n = GetLevelNumberByPassword(password);
            if (n == -1)
            {
                return null;
            }

            return GetLevel(n);
        }

        public abstract GameLevel GetLevel(int n);

        protected internal abstract int GetLevelNumberByPassword(string pass);
    }
}